File: apps/frontend/lib/workflow/evaluator/fieldAccessors.ts
import type { WorkflowItem } from '@/lib/types/workflow'
import { AI_FIELD_PREFIX } from '@/lib/workflow/registry/conditionFields'

export interface FieldLookup {
  found: boolean
  value: unknown
}

const NOT_FOUND: FieldLookup = { found: false, value: undefined }

export interface AccessorContext {
  depth: number
}

export function getFieldValue(item: WorkflowItem, field: string, ctx: AccessorContext): FieldLookup {
  if (field.startsWith(AI_FIELD_PREFIX)) {
    const key = field.slice(AI_FIELD_PREFIX.length)
    if (!key || !item.ai || !(key in item.ai)) return NOT_FOUND
    return { found: true, value: item.ai[key] }
  }

  switch (field) {
    case 'type': return { found: true, value: item.type }
    case 'name': return { found: true, value: item.name }
    case 'extension':
      return item.extension === undefined ? NOT_FOUND : { found: true, value: item.extension }
    case 'mime_type':
      return item.mimeType === undefined ? NOT_FOUND : { found: true, value: item.mimeType }
    case 'path': return { found: true, value: item.path }
    case 'size':
      if (item.type === 'folder') return { found: true, value: recursiveFolderSize(item) }
      return { found: true, value: item.stat.size }
    case 'created_at': return { found: true, value: item.stat.createdAt }
    case 'modified_at': return { found: true, value: item.stat.modifiedAt }
    case 'accessed_at': return { found: true, value: item.stat.accessedAt }
    case 'is_hidden': return { found: true, value: item.flags.hidden }
    case 'is_executable': return { found: true, value: item.flags.executable }
    case 'is_readable': return { found: true, value: item.flags.readable }
    case 'is_writable': return { found: true, value: item.flags.writable }
    case 'is_empty':
      if (item.type !== 'folder') return NOT_FOUND
      return { found: true, value: isFolderEmpty(item) }
    case 'children_count':
      return { found: true, value: item.children?.length ?? 0 }
    case 'depth':
      return { found: true, value: ctx.depth }
    default:
      return NOT_FOUND
  }
}

function recursiveFolderSize(item: WorkflowItem): number {
  let total = item.stat.size ?? 0
  for (const child of item.children ?? []) {
    total += child.type === 'folder' ? recursiveFolderSize(child) : (child.stat.size ?? 0)
  }
  return total
}

function isFolderEmpty(item: WorkflowItem): boolean {
  const children = item.children ?? []
  if (children.length === 0) return true
  return children.every((c) => c.type === 'folder' && isFolderEmpty(c))
}

File: apps/backend/app/modules/workflows/application/conditions/field_accessors.py
"""Resolve a condition ``field`` to a value on a WorkflowItem.

Fields are resolved through an explicit map (never dynamic attribute access) so a condition
can only read whitelisted item data. Mirrors the frontend accessor
(lib/workflow/evaluator/fieldAccessors.ts), with two deliberate differences:

* backend item type ``directory`` is exposed as ``folder`` to match the frontend vocabulary,
* ``ai.*`` fields are never present yet (no AI execution stage), so they report "not found"
  and fall through to the node's missing-field strategy.
"""

from typing import Any

from app.modules.workflows.domain.models import WorkflowItem

AI_FIELD_PREFIX = "ai."

# Sentinel distinguishing "field resolved to None" from "field absent".
_NOT_FOUND = object()


def get_field_value(item: WorkflowItem, field: str) -> Any:
    """Return the field value, or ``_NOT_FOUND`` when the field is absent."""
    if field.startswith(AI_FIELD_PREFIX):
        return _NOT_FOUND

    resolver = _RESOLVERS.get(field)
    if resolver is None:
        return _NOT_FOUND
    return resolver(item)


def is_found(value: Any) -> bool:
    return value is not _NOT_FOUND


def _resolve_type(item: WorkflowItem) -> Any:
    return "folder" if item.type == "directory" else "file"


def _resolve_extension(item: WorkflowItem) -> Any:
    return _NOT_FOUND if item.extension is None else item.extension


def _resolve_mime_type(item: WorkflowItem) -> Any:
    return _NOT_FOUND if item.mime_type is None else item.mime_type


def _resolve_is_empty(item: WorkflowItem) -> Any:
    return item.is_empty if item.type == "directory" else _NOT_FOUND


_RESOLVERS = {
    "type": _resolve_type,
    "name": lambda item: item.name,
    "extension": _resolve_extension,
    "mime_type": _resolve_mime_type,
    "path": lambda item: item.path,
    "size": lambda item: item.size if item.size is not None else 0,
    "created_at": lambda item: item.created_at,
    "modified_at": lambda item: item.modified_at,
    "accessed_at": lambda item: item.accessed_at,
    "is_hidden": lambda item: item.is_hidden,
    "is_executable": lambda item: item.is_executable,
    "is_readable": lambda item: item.is_readable,
    "is_writable": lambda item: item.is_writable,
    "is_empty": _resolve_is_empty,
    "children_count": lambda item: item.children_count,
    "depth": lambda item: item.depth,
}
